File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/play_relationships.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"
)

// playRelationship is a directed relationship edge between two campaign
// entities (campaign member character ids or npc ids).
type playRelationship struct {
	SourceID string
	TargetID string
	Kind     string
	Score    int
}

func playRelationshipResponse(rel *playRelationship) map[string]interface{} {
	return map[string]interface{}{
		"source_id": rel.SourceID,
		"target_id": rel.TargetID,
		"kind":      rel.Kind,
		"score":     rel.Score,
	}
}

// isPlayCampaignEntity reports whether id names an existing campaign entity:
// a member's character id or an npc id.
func isPlayCampaignEntity(c *playCampaign, id string) bool {
	for _, m := range c.Members {
		if m.CharacterID == id {
			return true
		}
	}
	if c.NPCs != nil {
		if _, ok := c.NPCs[id]; ok {
			return true
		}
	}
	return false
}

// handlePlayCampaignRelationshipSub routes the "relationships" and
// "relationships/..." sub-paths of a play campaign. It returns false if rest
// does not name a relationships path, so the caller can fall through to its
// own routing.
func handlePlayCampaignRelationshipSub(w http.ResponseWriter, r *http.Request, campaignID, rest string) bool {
	if rest == "relationships" {
		switch r.Method {
		case http.MethodPost:
			handleCreatePlayRelationship(w, r, campaignID)
		case http.MethodGet:
			handleListPlayRelationships(w, r, campaignID)
		default:
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		}
		return true
	}
	if !strings.HasPrefix(rest, "relationships/") {
		return false
	}
	relRest := strings.TrimPrefix(rest, "relationships/")
	parts := strings.SplitN(relRest, "/", 3)
	if len(parts) != 3 || parts[0] == "" || parts[1] == "" || parts[2] == "" {
		return false
	}
	if r.Method != http.MethodPut {
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		return true
	}
	handleUpdatePlayRelationship(w, r, campaignID, parts[0], parts[1], parts[2])
	return true
}

// handleCreatePlayRelationship lets the campaign dm create a new directed
// relationship edge between two campaign entities.
func handleCreatePlayRelationship(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		SourceID string   `json:"source_id"`
		TargetID string   `json:"target_id"`
		Kind     string   `json:"kind"`
		Score    *float64 `json:"score"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.SourceID == "" || req.TargetID == "" || req.Kind == "" || req.Score == nil {
		writeError(w, http.StatusBadRequest, "source_id, target_id, kind, and score are required")
		return
	}
	if req.SourceID == req.TargetID {
		writeError(w, http.StatusBadRequest, "source_id and target_id must differ")
		return
	}
	score := *req.Score
	if score != float64(int(score)) || score < -100 || score > 100 {
		writeError(w, http.StatusBadRequest, "score must be an integer from -100 through 100")
		return
	}
	scoreInt := int(score)

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the dm may create relationship edges")
		return
	}
	if !isPlayCampaignEntity(c, req.SourceID) || !isPlayCampaignEntity(c, req.TargetID) {
		playMu.Unlock()
		writeError(w, http.StatusNotFound, "unknown campaign entity")
		return
	}
	for _, rel := range c.Relationships {
		if rel.SourceID == req.SourceID && rel.TargetID == req.TargetID && rel.Kind == req.Kind {
			playMu.Unlock()
			writeError(w, http.StatusConflict, "relationship edge already exists")
			return
		}
	}

	rel := &playRelationship{
		SourceID: req.SourceID,
		TargetID: req.TargetID,
		Kind:     req.Kind,
		Score:    scoreInt,
	}
	c.Relationships = append(c.Relationships, rel)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusCreated, playRelationshipResponse(rel))
}

// handleUpdatePlayRelationship lets the campaign dm update the score of an
// existing relationship edge.
func handleUpdatePlayRelationship(w http.ResponseWriter, r *http.Request, campaignID, sourceID, targetID, kind string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		Score *float64 `json:"score"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Score == nil {
		writeError(w, http.StatusBadRequest, "score is required")
		return
	}
	score := *req.Score
	if score != float64(int(score)) || score < -100 || score > 100 {
		writeError(w, http.StatusBadRequest, "score must be an integer from -100 through 100")
		return
	}
	scoreInt := int(score)

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the dm may update relationship edges")
		return
	}
	var rel *playRelationship
	for _, candidate := range c.Relationships {
		if candidate.SourceID == sourceID && candidate.TargetID == targetID && candidate.Kind == kind {
			rel = candidate
			break
		}
	}
	if rel == nil {
		playMu.Unlock()
		writeError(w, http.StatusNotFound, "relationship edge not found")
		return
	}

	rel.Score = scoreInt
	resp := playRelationshipResponse(rel)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusOK, resp)
}

// handleListPlayRelationships returns all relationship edges for a campaign
// in insertion order. Any authenticated campaign member may call this.
func handleListPlayRelationships(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if !isPlayMember(c, username) {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the owner or a member may view relationships")
		return
	}
	edges := make([]map[string]interface{}, 0, len(c.Relationships))
	for _, rel := range c.Relationships {
		edges = append(edges, playRelationshipResponse(rel))
	}
	playMu.Unlock()

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"edges": edges,
	})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/play_relationships.rb
# frozen_string_literal: true

# Route definitions for play-campaign relationship graph edges.

# --- Create relationship edge ---
post '/v1/play/campaigns/:id/relationships' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  username = require_dm_actor!
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  body = parse_json_body
  validate_relationship_body!(body)

  source_id = body['source_id']
  target_id = body['target_id']
  kind = body['kind']
  score = body['score']

  json_error(400, 'self-edges are not allowed') if source_id == target_id

  unless Storage.play_campaign_entity_exists?(campaign_id, source_id) && Storage.play_campaign_entity_exists?(campaign_id, target_id)
    json_error(404, 'entity not found')
  end

  if Storage.play_campaign_relationship_exists?(campaign_id, source_id, target_id, kind)
    json_error(409, 'relationship already exists')
  end

  Storage.create_play_campaign_relationship(campaign_id, source_id, target_id, kind, score)

  status 201
  JSON.dump(
    source_id: source_id,
    target_id: target_id,
    kind: kind,
    score: score
  )
end

# --- Update relationship edge ---
put '/v1/play/campaigns/:id/relationships/:source_id/:target_id/:kind' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  username = require_dm_actor!
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  body = parse_json_body
  score = body['score']
  unless score.is_a?(Integer) && (-100..100).cover?(score)
    json_error(400, 'invalid score')
  end

  relationship = Storage.load_play_campaign_relationship(campaign_id, params[:source_id], params[:target_id], params[:kind])
  json_error(404, 'relationship not found') unless relationship

  Storage.update_play_campaign_relationship(campaign_id, relationship[:source_id], relationship[:target_id], relationship[:kind], score)

  JSON.dump(
    source_id: relationship[:source_id],
    target_id: relationship[:target_id],
    kind: relationship[:kind],
    score: score
  )
end

# --- List relationship edges ---
get '/v1/play/campaigns/:id/relationships' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  user = authenticate_actor!
  username = user[:username]

  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_access!(campaign, username)

  edges = Storage.load_play_campaign_relationships(campaign_id)

  JSON.dump(edges: edges)
end
